src/main/java/com/google/sps/servlets/GmailServlet.java

```diff
@@ -66,15 +74,131 @@ public void doGet(
+   *
+   * @param nDays number of days to check the user's account for
+   * @param gmailClient Gmail service with valid google credential
+   * @return number of unread, important emails from last nDays days
+   * @throws IOException if an issue occurs with the Gmail Service
+   */
+  private int getImportantUnreadEmailCount(int nDays, GmailClient gmailClient) throws IOException {
+    String searchQuery = GmailClient.emailQueryString(nDays, "d", true, true, "");
+    return gmailClient.listUserMessages(searchQuery).size();
+  }
+
+  /**
+   * Get the sender of the most unread emails from the last n days in a user's Gmail account
+   *
+   * @param nDays number of days to check the user's account for
+   * @param gmailClient Gmail service with valid google credential
+   * @return sender of the most unread emails from the last nDays days
+   * @throws IOException if an issue occurs with the Gmail Service
+   */
+  private String getMostFrequentSender(int nDays, GmailClient gmailClient) throws IOException {
+    // Get user's messages - if none present, return "" as there is no sender to extract
+    String searchQuery = GmailClient.emailQueryString(nDays, "d", true, false, "");
+    GmailClient.MessageFormat messageFormat = GmailClient.MessageFormat.METADATA;
+    List<Message> unreadEmails;
+    unreadEmails =
+        gmailClient.listUserMessages(searchQuery).stream()
+            .map(
+                (Message m) -> {
+                  try {
+                    return gmailClient.getUserMessage(m.getId(), messageFormat);
+                  } catch (IOException e) {
+                    throw new RuntimeException(e);
+                  }
+                })
+            .collect(Collectors.toList());
+    if (unreadEmails.isEmpty()) {
+      return "";
+    }
+
+    // Find out which sender sent the most emails
+    // Extracted from headers within messages.
+    // Headers have the name "From" and two types of values:
+    // <sampleemail@sample.com> (if name is not available)
+    // OR
+    // Sample Sender <sampleemail@sample.com>
+    // If a name is available, this is extracted. Otherwise, the email is extracted
+    HashMap<String, Integer> senders = new HashMap<>();
+    unreadEmails.forEach(
+        (Message m) -> {
+          List<MessagePartHeader> senderList = m.getPayload().getHeaders();
+          senderList =
+              senderList.stream()
+                  .filter((MessagePartHeader header) -> header.getName().equals("From"))
+                  .collect(Collectors.toList());
+
+          String sender = senderList.get(0).getValue();
+
+          senders.put(sender, senders.get(sender) != null ? senders.get(sender) + 1 : 1);
+        });
+    String sender = senders.entrySet().stream().max(Map.Entry.comparingByValue()).get().getKey();
+
+    if (sender.charAt(0) == '<') {
+      return sender.substring(1, sender.length() - 1);
+    } else {
+      return sender.split("<")[0].trim();
```

Comment:
Perhaps a comment to show how `sender` is supposed to typically look, hard to imagine what these operations would do to `sender`.

Reply:
There are comments regarding this on lines 176 to 182. I changed the variable name to `headerValue` so the connection between this variable and that comment is a bit more intuitive.